devdoc/waiter.md

```diff
@@ -0,0 +1,231 @@
+
+#ifdef __cplusplus
+}
+#endif /* __cplusplus */
+
+```
+
+### waiter_create
+```c
+    MOCKABLE_FUNCTION(, WAITER_HANDLE, waiter_create)
+```
+
+`waiter_create` creates the waiter and returns it.
+
+**SRS_WAITER_43_001: [** `waiter_create` shall call `THANDLE_MALLOC` with `dispose` as `waiter_dispose`. **]**
+
+**SRS_WAITER_43_002: [** If there are any failures, `waiter_create` shall fail and return `NULL`. **]**
+
+### waiter_dispose
+```c
+    void waiter_dispose(WAITER* waiter)
+```
+
+**SRS_WAITER_43_003: [** `waiter_dispose` disposes the given `waiter`. **]**
+
+**SRS_WAITER_43_004: [** `waiter_dispose` shall obtain the current state of the waiter by calling `interlocked_and`. **]**
+
+**SRS_WAITER_43_005: [** If the `REGISTERED` bit is set, `waiter_dispose` shall call the `notification_callback` given to `waiter_register_notification` with `context` as the `context` given to `waiter_register_notification`, `data` as `NULL` and `result` as `WAITER_RESULT_ABANDONED`. **]**
+
+**SRS_WAITER_43_006: [** If the `NOTIFIED` bit is set, `waiter_dispose` shall: **]**
+
+- **SRS_WAITER_43_044: [** call the `notify_complete_callback` given to `waiter_notify` with `context` as the `context` given to `waiter_notify`, and `result` as `WAITER_RESULT_ABANDONED`. **]**
+
+- **SRS_WAITER_43_045: [** release the reference to `data` stored in `waiter` by calling `THANDLE_ASSIGN`. **]**
+
+### waiter_register_notification
+```c
+    MOCKABLE_FUNCTION(, WAITER_RESULT, waiter_register_notification, THANDLE(WAITER), waiter, NOTIFICATION_CALLBACK, notification_callback, void*, context, THANDLE(ASYNC_OP)*, out_op);
+```
+
+`waiter_register_notification` registers the given `notification_callback` to be called when `waiter_notify` is called.
+
+**SRS_WAITER_43_007: [** If `waiter` is `NULL`, `waiter_register_notification` shall fail and return `WAITER_RESULT_ERROR`. **]**
+
+**SRS_WAITER_43_008: [** If `notification_callback` is `NULL`, `waiter_register_notification` shall fail and return `WAITER_RESULT_ERROR`. **]**
+
+**SRS_WAITER_43_009: [** If `out_op` is `NULL`, `waiter_register_notification` shall fail and return `WAITER_RESULT_ERROR`. **]**
+
+**SRS_WAITER_43_010: [** `waiter_register_notification` shall call `interlocked_or` to set the `REGISTERED` bit and obtain the current state of the `waiter`. **]**
+
+**SRS_WAITER_43_011: [** If the `REGISTERED` bit was already set in the current state, `waiter_register_notification` shall fail and return `WAITER_RESULT_REFUSED`. **]**
+
+**SRS_WAITER_43_012: [** If the `NOTIFIED` bit is set, `waiter_register_notification` shall : **]**
+
+- **SRS_WAITER_43_042: [** call `interlocked_exchange` to increment the `EPOCH` and unset the `NOTIFIED` bit and `REGISTERED` bit. **]**
+
+- **SRS_WAITER_43_013: [** call the given `notification_callback` with `context` as the given `context`, `data` as the `data` that was given to `waiter_notify` and `result` as `WAITER_RESULT_OK`. **]**
+
```

Comment:
Here is the problem:
once the epoch was changed and the bits were reset, the data that is stored is not valid anymore (the next microsecond after the .SRS_WAITER_43_042 spec was executed, another call to notify comes in and stores the data overwriting your existing data.

The only way I believe you can avoid this is by having barriers for putting and taking the data out of the waiter.

Think the following states:
- NO_DATA
- NOTIFYING - stores notify data (the data, etc.)
- NOTIFIED
- REGISTERING - stores register_notification data (callback)
- REGISTERED
- NOTIFIED_REGISTERING - takes out data preparing to call  notification_callback
- REGISTER_NOTIFYING - takes out data preparing to call notification_callback

The *ING states should be short and always be treated as a lock.

--------

An alternative would be to use a mutex/lock

----------

Another alternative solution (maybe smarter but more complex) would be to bundle everything in a structure and use a pointer to make sure that you atomically swap the state and the data.

On Windows (and others), due to malloc alignment you can use the least significant bits to store some state (pretty sure you can use 2 bits there to store the states you want.